File: Algorithms/src/1937. Maximum Number of Points with Cost/Code/solutions.java
class Solution {
    public long maxPoints(int[][] points) {
        // Get the number of rows (m) and columns (n) from the points array
        int m = points.length;
        int n = points[0].length;

        // Create a dp array to store the maximum points for each column in the current
        // row
        long[] dp = new long[n];

        // Initialize dp with the values from the first row
        for (int j = 0; j < n; ++j) {
            dp[j] = points[0][j];
        }

        // Iterate over each row starting from the second one
        for (int i = 1; i < m; ++i) {
            // Arrays to store the maximum values while traversing from left to right and
            // right to left
            long[] leftMax = new long[n];
            long[] rightMax = new long[n];
            // Array to store the new dp values for the current row
            long[] newDp = new long[n];

            // Calculate the maximum points that can be obtained by moving from left to
            // right
            leftMax[0] = dp[0]; // Initialize the first element
            for (int j = 1; j < n; ++j) {
                // Maximum of the previous leftMax value or the current dp value plus the column
                // index
                leftMax[j] = Math.max(leftMax[j - 1], dp[j] + j);
            }

            // Calculate the maximum points that can be obtained by moving from right to
            // left
            rightMax[n - 1] = dp[n - 1] - (n - 1); // Initialize the last element
            for (int j = n - 2; j >= 0; --j) {
                // Maximum of the next rightMax value or the current dp value minus the column
                // index
                rightMax[j] = Math.max(rightMax[j + 1], dp[j] - j);
            }

            // Calculate the dp values for the current row
            for (int j = 0; j < n; ++j) {
                // The new dp value is the maximum of leftMax[j] - j or rightMax[j] + j plus the
                // points at the current position
                newDp[j] = Math.max(leftMax[j] - j, rightMax[j] + j) + points[i][j];
            }

            // Update dp to the new values calculated for the current row
            dp = newDp;
        }

        // Find the maximum value in the dp array after processing all rows
        long maxPoints = dp[0];
        for (int j = 1; j < n; ++j) {
            maxPoints = Math.max(maxPoints, dp[j]);
        }

        // Return the maximum points that can be obtained
        return maxPoints;
    }
}

File: Algorithms/src/1937. Maximum Number of Points with Cost/Code/solutions.cpp
class Solution
{
public:
    long long maxPoints(vector<vector<int>> &points)
    {
        int m = points.size();    // Number of rows
        int n = points[0].size(); // Number of columns

        // DP array to store the maximum points that can be collected up to the current row
        vector<long long> dp(n, 0);

        // Initialize dp array with the values from the first row
        for (int j = 0; j < n; ++j)
        {
            dp[j] = points[0][j];
        }

        // Traverse through each row starting from the second row
        for (int i = 1; i < m; ++i)
        {
            // Arrays to store the maximum values moving from left to right and right to left
            vector<long long> leftMax(n), rightMax(n), newDp(n);

            // Calculate leftMax array
            // leftMax[j] will store the maximum value for positions [0, j] considering the left side
            leftMax[0] = dp[0];
            for (int j = 1; j < n; ++j)
            {
                leftMax[j] = max(leftMax[j - 1], dp[j] + j);
            }

            // Calculate rightMax array
            // rightMax[j] will store the maximum value for positions [j, n-1] considering the right side
            rightMax[n - 1] = dp[n - 1] - (n - 1);
            for (int j = n - 2; j >= 0; --j)
            {
                rightMax[j] = max(rightMax[j + 1], dp[j] - j);
            }

            // Calculate the new DP values for the current row
            for (int j = 0; j < n; ++j)
            {
                // The new DP value for the current cell is calculated by taking the maximum of
                // either the leftMax or rightMax adjusted by the current column index,
                // and then adding the current cell's points
                newDp[j] = max(leftMax[j] - j, rightMax[j] + j) + points[i][j];
            }

            // Update the dp array with the newly calculated values for the next row
            dp = newDp;
        }

        // The answer will be the maximum value in the dp array after processing all rows
        return *max_element(dp.begin(), dp.end());
    }
};
